src/compiler/checker.ts

```diff
@@ -30279,7 +30307,16 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             else if (t.flags & TypeFlags.StructuredType) {
                 const prop = getPropertyOfType(t, name);
                 if (prop) {
-                    return isCircularMappedProperty(prop) ? undefined : removeMissingType(getTypeOfSymbol(prop), !!(prop && prop.flags & SymbolFlags.Optional));
+                    if (isCircularMappedProperty(prop)) {
+                        return undefined;
+                    }
+                    const type = getTypeOfSymbol(prop);
+                    if (type !== nonInferrableUnknownType) {

```

Comment:
Since `nonInferrableUnknownType` doesn't hold any useful information for contextual typing I let it fallthrough to index signatures here. Note that `{ x: nonInferrableUnknownType } & { x: string }` **shouldn't** return `nonInferrableUnknownType` for `'x'` property here. So if the constraint already defines a property then index signatures won't be consulted.